observable.go

```diff
@@ -1111,6 +1112,69 @@ func (o *observable) Repeat(count int64, frequency Duration) Observable {
 	return newColdObservableFromFunction(f)
 }
 
+// Sample returns an Observable that emits the most recent items emitted by the source
+// ObservableSource whenever the input Observable emits an item.
+func (o *observable) Sample(obs Observable) Observable {
+	f := func(out chan interface{}) {
+		mainChan := make(chan interface{})
+		obsChan := make(chan interface{})
+		var lastEmittedItem interface{}
+		isAnyItemEmitted := false
+		ctx, cancel := context.WithCancel(context.Background())
+
+		go func() {
+			it := o.iterable.Iterator(ctx)
+			for {
+				if item, err := it.Next(ctx); err == nil {
+					mainChan <- item
+				} else {
+					break
+				}
+			}
+			close(mainChan)
+		}()
+
+		go func() {
+			it := obs.Iterator(ctx)
+			for {
+				if item, err := it.Next(ctx); err == nil {
+					obsChan <- item
+				} else {
+					break
+				}
+			}
+			close(obsChan)
+		}()
+
+	mainLoop:
+		for {
+			select {
+			case item, ok := <-mainChan:
+				if ok {
+					lastEmittedItem = item
+					isAnyItemEmitted = true
+				} else {
+					cancel()
+					close(out)
+					break mainLoop
+				}
+			case _, ok := <-obsChan:
+				if ok {
+					if isAnyItemEmitted {
+						out <- lastEmittedItem
```

Comment:
Shouldn't we put `lastEmittedItem` to true then?

According to the documentation, the following code should print only one _1_:
```go
ch := make(chan interface{}, 1)
obs := FromChannel(ch)
sample := obs.Sample(Interval(make(chan struct{}), time.Second))
sample.Subscribe(handlers.NextFunc(func(i interface{}) {
	fmt.Printf("%v\n", i)
}))
ch <- 1
```

Here, though, it keeps printing 1.

Reply:
You're right. I have not tested such situation. I've just pushed changes that include fixed solution and tests.